### PR TITLE
Enabling showon to work in list options in JForm (resurrected)

### PR DESCRIPTION
**Pull request by GeraintEdwards**
(Resurrection of old Pull Request https://github.com/joomla/joomla-cms/pull/17212  where the fork/branch got royally messed up)

### Summary of Changes
The showon attribute currently works for fields in Jforms to allow their display to be conditional on the value of another field.  This PR makes some simple changes to enable this functionality to work on list option value too - in other works the display of a specific list option can now be dependent on the value of another field.

The changes required to make this work are minimal:
1.  In libraries/joomla/form/fields/list.php we add pass new attribute name as an option 

`$listoptions['option.attr'] = 'optionattr';`

this means that the list renderer will use the option.attr value we pass in to add additional attributes to the option.

2. In the getOptions method of  libraries/joomla/form/fields/list.php  we look for a 'showon' attribute and if its specified pass it as 'optionattr'

3.  In media/jui/js/cms-uncompressed.js/media/jui/js/cms.js we have to disable the normal animation effect for option values (this does work in Chrome but not in other browsers) and add some special handling to toggle the display of the option and trigger an update in the chosen display where needed).

So all in all minimal changes 

### Testing Instructions

Best test is to edit an existing form e.g. administrator/components/com_content/config.xml and look for the declaration of targetb and add a showon attribute e.g.

`<option value="1" showon="targeta:1">JBROWSERTARGET_NEW</option>`

This make the browser new option in the URL b target drop down dependent on the value of URL a target drop down. 

### Expected result

When you edit the articles options and  set URL A target drop down equal to 'new browser window' this option is available for URL B but otherwise its hidden.

optionvisible

optionhiden

### Actual result

As explained above.

### CODE REVIEW

**Comment by GeraintEdwards** [comment]
@C-Lodder @dgt41 please continue the discussion of the previous PR here (sorry I messed up my local branch/fork and had to start again :( )

> The showon code needs a lot of work:
> 
>     It really slow using jQuery and a large form (I have a prototype version implemented using native query selectors with a pseudo polyfill that falls back to jQuery where the query selector methods don't exist)
>     the toggle needs to be replaced with a CSS transition (this is a very small performance benefit compared to the gains available from improving the finding of conditional elements and processing them on large forms)
>     the toggle/its replacement needs to respect the original CSS state - e.g. a div shouldn't be reset to block if it was an inline-block beforehand
>     the xml syntax should allow for and/or to be combined intelligently with appropriate associative ordering/parenthesis support
> 
> A lot of this could be done before 4.0 and I propose putting a PR together to deal with it in the next couple of weeks.
> 
> In the meantime cane we move forward with this PR please?

File: libraries/joomla/form/fields/list.php [hunk 1]
```diff
@@ -157,6 +166,14 @@ protected function getOptions()
 			$tmp['onclick']  = (string) $option['onclick'];
 			$tmp['onchange'] = (string) $option['onchange'];
 
+			if ((string) $option['showon'])
+			{
+				$tmp['optionattr']   = " data-showon = '" .
```
Comment (Quy) [inline libraries/joomla/form/fields/list.php]: Remove extra spaces before the first `=` and remove space before and after the second `=`.

File: media/jui/js/cms-uncompressed.js [hunk 2]
```diff
@@ -140,10 +138,20 @@ if (typeof(Joomla) === 'undefined') {
 			}
 
 			// If conditions are satisfied show the target field(s), else hide
-			if (animate) {
+			// note that animations don't work on list options other than in Chrome
```
Comment (Quy) [inline media/jui/js/cms-uncompressed.js]: Capitalize `N` in `note`, otherwise, it appears to be one sentence.

**Comment by GeraintEdwards** [comment]
Can we get this merged please - it adds a useful configuration feature to JForm with minimal code changes

**Comment by brianteeman** [comment]
Like all pr it needs testers

**Comment by tonypartridge** [comment]
@Quy @ggppdk could you guys pretty please run some tests? :-) It would be hugely appreciated!

**Comment by ggppdk** [comment]
@tonypartridge  @GeraintEdwards 

I remember this PR
This will be a nice feature for both configuration and content forms, it will be useful for several core and 3rd party extensions, it will give a better UX

I ll try to test by Monday if it does not get enough tests by then

**Comment by tonypartridge** [comment]
Thank you!!

**Comment by tonypartridge** [comment]
I have tested this item :white_check_mark: successfully on 9cc2b07c55b0a681e3ef1518fe9ef3d52cff6042<br /><br />Tested, working well. Don't see any issues at all.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18998">issues.joomla.org/tracker/joomla-cms/18998</a>.</sub>

**Comment by carcam** [comment]
I have tested this item :white_check_mark: successfully on 9cc2b07c55b0a681e3ef1518fe9ef3d52cff6042<br /><br />Tested correctly with different advanced combinations of showon:

Form File: *administrator/components/com_content/config.xml*

## Test proposed by @GeraintEdwards

Success!!

## Test OR syntax
Testing code(line 498):

```
<option value="1" showon="targeta:1[OR]targeta:2">JBROWSERTARGET_NEW</option>
```

Success!!

## Testing AND syntax
Testing Code (Line 513):

```
<option value="1" showon="targeta:1[AND]targetb:1">JBROWSERTARGET_NEW</option>
```

Success!!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18998">issues.joomla.org/tracker/joomla-cms/18998</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by ReLater** [comment]
FYI: Doc added: https://docs.joomla.org/index.php?title=List_form_field_type#Showon_attribute_for_list_options

**Pull request merged**